stdlib/fun.mli

```diff
@@ -31,3 +31,25 @@ val flip : ('a -> 'b -> 'c) -> ('b -> 'a -> 'c)
 val negate : ('a -> bool) -> ('a -> bool)
 (** [negate p] is the negation of the predicate function [p]. For any
     argument [x], [(negate p) x] is [not (p x)]. *)
+
+val protect : finally:(unit -> unit) -> (unit -> 'a) -> 'a
+(** [protect ~finally work] invokes [work ()] and then [finally ()]
+    before [work ()] returns with its value or an exception. In the
+    latter case the exception is re-raised after [finally ()]. If
+    [finally ()] raises an exception, then the exception
+    {!Finally_raised} is raised instead.
+
+    [protect] can be used to enforce local invariants whether the
+    function returns normally or raises an exception. However, it does
+    not protect against unexpected exceptions raised inside [finally
+    ()] such as Out_of_memory, Stack_overflow, or asynchronous
+    exceptions raised by signal handlers (e.g. Sys.Break). *)
+
+exception Finally_raised of exn
+(** Raised by [protect ~finally work] when [finally] raises an
+    exception. The argument is the exception that was raised.
+    Finally_raised can contain or shadow an unexpected exception raised
+    in [finally] or [work], such as Invalid_argument, Out_of_memory,
+    Stack_overflow, or any asynchronous exception raised by finalisers
+    or signal handlers (e.g. Sys.Break); it should be treated similarly
+    to those. *)
```

Comment:
It may not be exactly clear to the reader how the other ones are supposed to be treated. I would be explicit in saying that it is not meant to be caught except maybe at the toplevel of your program.